res/controllers/Reloop-Mixage.scripts.js

```diff
@@ -0,0 +1,784 @@
+            // make connections for the fx on LEDs
+            const fxGroup = "group_"+channel+"_enable";
+            Mixage.fxOnConnection.push(engine.makeConnection("[EffectRack1_EffectUnit"+effectUnit+"]", fxGroup, function() { Mixage.toggleFxLED(channel); }));
+
+            // set soft takeovers for effectunit meta
+            engine.softTakeover("[EffectRack1_EffectUnit"+effectUnit+"]", "super1", true);
+            engine.setValue("[EffectRack1_EffectUnit"+effectUnit+"]", "show_focus", 1);
+        }
+
+        // set soft takeover for filter effect
+        engine.softTakeover("[QuickEffectRack1_"+channel+"]", "super1", true);
+
+        // make connections for status LEDs
+        Mixage.vuMeterConnection.push(engine.makeConnection(channel, "vu_meter", function(val) { midi.sendShortMsg(0x90, Mixage.ledMap[channel].vu_meter, val * 7); }));
+        Mixage.loopConnection.push(engine.makeConnection(channel, "track_loaded", function() { Mixage.toggleReloopLED(channel); }));
+        Mixage.fxSelectConnection.push(engine.makeConnection("[EffectRack1_EffectUnit"+deck+"]", "focused_effect", function(value) { Mixage.handleFxSelect(value, channel); }));
+
+        // get current status and set LEDs accordingly
+        Mixage.toggleFxLED(channel);
+        Mixage.handleFxSelect(engine.getValue("[EffectRack1_EffectUnit"+deck+"]", "focused_effect"), channel);
+    });
+};
+
+Mixage.shutdown = function() {
+
+    // Disconnect all engine connections that are present
+    Mixage.vuMeterConnection.forEach(function(connection) { connection.disconnect(); });
+    Mixage.loopConnection.forEach(function(connection) { connection.disconnect(); });
+    Mixage.fxSelectConnection.forEach(function(connection) { connection.disconnect(); });
+    Mixage.fxOnConnection.forEach(function(connection) { connection.disconnect(); });
+
+    // Disconnect all controls from functions
+    Mixage.channels.forEach(function(channel) { Mixage.connectControlsToFunctions(channel, true); });
+
+    // all button LEDs off
+    for (let i = 0; i < 255; i++) {
+        midi.sendShortMsg(0x90, i, 0);
+    }
+};
+
+// Toggle the LED on the MIDI controller by sending a MIDI message
+Mixage.toggleLED = function(value, group, control) {
+    midi.sendShortMsg(0x90, Mixage.ledMap[group][control], value ? 0x7F : 0);
+};
+
+// Toggles the FX On LED / Off when no effect unit is activated for a channel / On when any effect unit is active for a channel
+Mixage.toggleFxLED = function(group) {
+    const fxChannel = "group_" + group + "_enable";
+    const enabledFxGroups = [];
+
+    for (let i = 1; i <= Mixage.numEffectUnits; i++) {
+        enabledFxGroups.push(engine.getValue("[EffectRack1_EffectUnit" + i + "]", fxChannel));
+    }
+
+    if (enabledFxGroups.indexOf(1) !== -1) {
+        Mixage.toggleLED(ON, group, "fx_on");
+    } else {
+        Mixage.toggleLED(OFF, group, "fx_on");
+    }
+};
+
+// Flash the Reloop LED if a loop is set but currently not active
```

Comment:
I think this needs discussion. LEDs should only really blink if they need the users attention, a loop that is set but not active is not uncommon and does not need urgent attention.

Reply:
Currently it blinks slowly (1 second on/off) to indicate that reloop is available for the deck. In my opinion the slow blinking rate doesn't draw immediate attention and is a helpfull indication that reloop is available. However this is open for debate. I chose this implementation so I could use the `loop/loop in` and `reloop/loop out` to communicate to a user if a loop is active (both leds on) or if only a loop in or loop out point has been set (although a loop out point cannot be set without a loop in point present)

Reply:
While your implementation is certainly valid and makes sense, I'd personally prefer if the behavior was simpler and would follow the mixxx UI more closely for the sake of consistency (the loop button should bind to `beatloop_activate` for input and `loop_active` for output and the reloop button bound to `reloop_toggle`). wdyt?

Reply:
Indicating whether reloop is actually available makes a ton of sense, but it might be better implemented in C++ so all mappings and the GUI can benefit from it.

Reply:
IMO binding the outputs as you suggested doesn't make optimal use of the LEDs on this particular controller. Effectively it means that the reloop LED is only lit when the reloop button is pressed communicating nothing about the state of the deck to the user. All LEDs on this controller (beside the -/+) button do communicate something about the state of the controller to the user. I even looked into using the -/+ LEDs to indicate the jog direction of a deck but I could not find a way to implement this in a clean manner. Buttons that only light up as long as they are being pressed is IMO a missed opportunity to communicate deck state. 

I do agree with you point that it might be better to implement this natively in C++ but I can always adjust this mapping accordingly when that happens.

Reply:
> I do agree with you point that it might be better to implement this natively in C++ but I can always adjust this mapping accordingly when that happens.

Would you be interested in implementing the C++ functionality?

Reply:
>Would you be interested in implementing that feature in C++?

Maybe, although I have no prior experience with C++ or any compiling programming language for that matter. I might look into it after this mapping has been merged. I'm curious if there is any documentation available to get started with C++ in Mixxx

Reply:
>Indicating whether reloop is actually available makes a ton of sense

So do you agree with the current reloop blinking implementation in the mapping? Or should be adjusted or removed before the pull request can be accepted? I think I made my case as to why I chose this implementation. In any case it is not as egregious as the `load` button flashing on beat which the controller [originally ](https://youtu.be/iXk5s0fQPIQ?si=0PHyUuqRO47YBw6-&t=105)shipped with

Reply:
I'm proficient in C++ and I've taken a look at the source, but I have honestly no clue where to even start. Also I don't know what @Swiftb0y you're suggesting. Change Mixxx behaviour or create new JS variable binding?

I'd suggest we resolve this by either accepting the functionality as-is or changing the script to use standard `loop_active`.

Reply:
`loop_active` should light up the loop button(s) IMO.
With the current implementation it would blink as soon as a track with a temporary loop is loaded?
I agree that turning the LED on only when the button is pressed is not helpful.

Personally I use blinking when the loop is active, it's an important temporary state that deserves attention IMO.
So it could be
* on: reloop available
* blink: loop is active

Reply:
In this mapping the LEDS for the `LOOP` / `LOOP IN` and `RELOOP` / `LOOP OUT`  buttons are controlled by a combination of the engine parameters `loop_enabled`,  `loop_start_position` and `loop_end_position` using the following logic

- If `loop_enabled` is true both the `LOOP` / `LOOP IN` and `RELOOP` / `LOOP OUT` light up
- If `loop_enabled` is false and `loop_start_position` and `loop_end_position` are set  the `RELOOP` / `LOOP OUT` button will flash slowly (1 second interval) indicating that reloop is available 
- If `loop_enabled` is false and `loop_start_position` and `loop_end_position` are not set  the ``LOOP` / `LOOP IN` and `RELOOP` / `LOOP OUT` will be turned off
- If `loop_enabled` is false and only the `loop_start_position` is set (occurs when manually setting a loop in point) the `LOOP` / `LOOP IN` will light up.
- If `loop_enabled` is false and only the `loop_end_position` is set (which is currently not possible in Mixxx) the `RELOOP` / `LOOP OUT` will light up

I'm not using `loop_active` as this will interfere with the LED feedback I have implemented for the loop adjustment mode. In this mode the user can adjust the loop in and out points with the jog wheel. The respective  `LOOP IN` or `LOOP OUT` button will flash rapidly (250ms interval) to indicate if the start, end or both are currently being adjusted with the jog wheel. I was inspired to add this feature after seeing this [youtube video](https://youtu.be/goyauO5rkv4?si=IItmDmU7hkeduu-z&t=202). 

I think the LED behaviour of the mapping is fine and follows the guidelines for Mixxx. I'm all for expanding the `reloop_toggle` output to indicate if a reloop is available. But until that feature is added I think it is fine to have it build into this mapping.

>Personally I use blinking when the loop is active, it's an important temporary state that deserves attention IMO. So it could be
>- on: reloop available
>- blink: loop is active

While I do see your point about it being a temporary state it is the exact opposite of the current Mixxx implementation. Where an enabled state is usually represented by a steady burning LED (the flashing Cue button on pause is a good example of this behavior). However I think either implementation could work to indicate reloop availability

Reply:
> If loop_enabled is false and loop_start_position and loop_end_position are set the RELOOP / LOOP OUT button will flash slowly (1 second interval) indicating that reloop is available

...which is true as soon as you load a track that had a loop set earlier (might even have been a short temporary loop_roll), which is most likely irrelevant in a new session.
Therefore I agree with @Swiftb0y (and the [mapping guideline](https://github.com/mixxxdj/mixxx/wiki/Contributing%20mappings#blinking-leds) btw) to **not** make it blink. This state does not deserve this much attention. You can just make it optional if you want that behavior for your own use case, but we shouldn't force it on anyone else.

Reply:
Thanks for clearing that up @ronso0 I removed the reloop blinking behaviour from the mapping